src/Microsoft.ML.Core/Data/ColumnType.cs

```diff
@@ -456,79 +456,29 @@ public override bool Equals(ColumnType other)
     /// </summary>
     public sealed class KeyType : PrimitiveType
     {
-        private KeyType(Type type, DataKind kind, ulong min, int count, bool contiguous)
+        public KeyType(Type type, ulong count)
             : base(type)
         {
             Contracts.AssertValue(type);
-            Contracts.Assert(kind.ToType() == type);
-
-            Contracts.CheckParam(min >= 0, nameof(min));
-            Contracts.CheckParam(count >= 0, nameof(count), "Must be non-negative.");
-            Contracts.CheckParam((ulong)count <= ulong.MaxValue - min, nameof(count));
-            Contracts.CheckParam((ulong)count <= kind.ToMaxInt(), nameof(count));
-            Contracts.CheckParam(contiguous || count == 0, nameof(count), "Must be 0 for non-contiguous");
-
-            Contiguous = contiguous;
-            Min = min;
+            if (count == 0 || type.ToMaxInt() < count)
+                throw Contracts.ExceptParam(nameof(count), "The cardinality of a {0} must not exceed {1}.MaxValue" +
+                    " and must be strictly positive.", nameof(KeyType), type.Name);
             Count = count;
         }
 
-        public KeyType(Type type, ulong min, int count, bool contiguous = true)
-            : this(type, CheckRefRawType(type), min, count, contiguous)
-        {
-        }
-
-        [BestFriend]
-        internal KeyType(DataKind kind, ulong min, int count, bool contiguous = true)
-            : this(ToRawType(kind), kind, min, count, contiguous)
-        {
-        }
-
-        private static DataKind CheckRefRawType(Type type)
-        {
-            Contracts.CheckValue(type, nameof(type));
-            Contracts.CheckParam(IsValidDataType(type), nameof(type));
-            var result = type.TryGetDataKind(out var kind);
-            Contracts.Assert(result);
-            return kind;
-
-        }
-
-        private static Type ToRawType(DataKind kind)
-        {
-            Contracts.CheckParam(IsValidDataKind(kind), nameof(kind));
-            return kind.ToType();
-        }
-
-        /// <summary>
-        /// Returns true iff the given DataKind is valid for a <see cref="KeyType"/>. The valid ones are
-        /// <see cref="DataKind.U1"/>, <see cref="DataKind.U2"/>, <see cref="DataKind.U4"/>, and <see cref="DataKind.U8"/>,
-        /// that is, the unsigned integer kinds.
-        /// </summary>
-        [BestFriend]
-        internal static bool IsValidDataKind(DataKind kind)
+        public KeyType(Type type, int count)
+            : this(type, (ulong)count)
         {
-            switch (kind)
-            {
-                case DataKind.U1:
-                case DataKind.U2:
-                case DataKind.U4:
-                case DataKind.U8:
-                    return true;
-                default:
-                    return false;
-            }
+            Contracts.CheckParam(0 < count, nameof(count), "The cardinality of a " + nameof(KeyType) + " must be strictly positive.");
```

Comment:
Do you need to check the two conditions in Line 472?

Reply:
This is checking that the input count is not negative (which would be nonsense when casting). The other two checks are run anyways as this constructor will call the first.